core/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java

```diff
@@ -0,0 +1,193 @@
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                responses.add(response);
+                latch.countDown();
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                responses.add(failure);
+                latch.countDown();
+            }
+        }).setBulkActions(1)
+                 // zero means that we're in the sync case, more means that we're in the async case
+                .setConcurrentRequests(randomIntBetween(0, 100))
+                .setBackoffPolicy(BackoffPolicy.exponentialBackoff())
+                .build();
+        indexDocs(bulkProcessor, numberOfAsyncOps);
+        latch.await(10, TimeUnit.SECONDS);
+        bulkProcessor.close();
+
+        assertThat(responses.size(), equalTo(numberOfAsyncOps));
+
+        // validate all responses
+        for (Object response : responses) {
+            if (response instanceof BulkResponse) {
+                BulkResponse bulkResponse = (BulkResponse) response;
+                for (BulkItemResponse bulkItemResponse : bulkResponse.getItems()) {
+                    if (bulkItemResponse.isFailed()) {
+                        BulkItemResponse.Failure failure = bulkItemResponse.getFailure();
+                        Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                        // we're not expecting that we still overwhelmed it even once
+                        if (rootCause instanceof EsRejectedExecutionException) {
+                            StringWriter sw = new StringWriter();
+                            failure.getCause().printStackTrace(new PrintWriter(sw));
+                            fail("Got unexpected reason [" + failure.getMessage() + "]: " + sw.toString());
+                        }
+                    }
+                }
+            } else {
+                Throwable t = (Throwable) response;
+                // we're not expecting any other errors
+                fail("Got unexpected failure: " + t);
+            }
+        }
+
+        // ensure that we can search for the new documents
+        client().admin().indices().prepareRefresh().execute().actionGet();
+
+        // validate we did not create any duplicates due to retries
+        SearchResponse results = client()
+                .prepareSearch("test")
+                .setTypes("test")
+                .setQuery(QueryBuilders.matchAllQuery())
+                .setSize(0)
+                .get();
+        assertThat(results.getHits().totalHits(), equalTo((long) numberOfAsyncOps));
+    }
+
+    private static void indexDocs(BulkProcessor processor, int numDocs) {
+        for (int i = 1; i <= numDocs; i++) {
+            processor.add(new IndexRequest("test", "test", Integer.toString(i)).source("field", randomRealisticUnicodeOfLengthBetween(1, 30)));
```

Comment:
`client().prepareIndex("test", "test", ....` is more common from what I've seen.

Not for this PR: it might be nice to have a `randomDoc(String index, String type, int id)` method that makes a random document. We use this `Integer.toString` idiom so so so much in the tests. The field being a realistic unicode string is a good touch too.